M2/Macaulay2/packages/UpdatePackage.m2

```diff
@@ -0,0 +1,122 @@
+    get("!"|"cd "|dir|/// && M2 --no-preload -e "uninstallPackage \"///|packagename|///\"" 2>&1 &///);
+    get("!"|"cd "|dir|/// && M2 --no-preload -e "installPackage(\"///|packagename|///\",FileName=>\"///|dir|packagename|".m2"|///\")" 2>&1 &///);
+    removeFile(dir|packagename|".m2"); 
+    if #listauxfiles > 0 then (
+        for f in listauxfiles do removeFile(dir|packagename|"/"|f); 
+        removeDirectory(dir|packagename);  
+    ); 
+    <<"--** package "<<packagename<<", version "<<v0<<", has been successfully installed **--"<<endl;
+    get("!"|///M2 --no-preload -e "loadPackage \"///|packagename|///\"; \"///|dir|///packageVersion.m2\"<<///|packagename|///.Options.Version<<close;" 2>&1 &///);
+    v := get(dir|"packageVersion.m2");
+    removeFile(dir|"packageVersion.m2");
+    if v != v0 then (
+        sf := (value packagename)#"source file";
+        err := "Error in loading the new version of the package "|packagename|". Check that in your path you do not have any older version of the package.";
+        i0 := 0; while fileExists(sf|"-"|toString(i0)) do i0 = i0+1;
+        try moveFile(sf,sf|"-"|toString(i0)) then (<<"--warning: the file "<<sf<<" has been renamed to "<<packagename|".m2-"|toString(i0)|" (you can also delete it)"<<endl) else error err;
+        get("!"|///M2 --no-preload -e "loadPackage \"///|packagename|///\"; \"///|dir|///packageVersion.m2\"<<///|packagename|///.Options.Version<<close;" 2>&1 &///);
+        v = get(dir|"packageVersion.m2");
+        removeFile(dir|"packageVersion.m2");
+        if v != v0 then error err;   
+    );    
+    return v;
+);
+updatePackageFromDevelopment String := packagename -> updatePackageFromDevelopment(packagename,{});
+
+auxFiles = packagename -> (
+    if not (value packagename).Options.AuxiliaryFiles then return {};
+    f := lines get("!ls "|(value packagename)#"auxiliary files");
+    if any(f,isDirectory) then error("this function is not yet available for the package "|packagename);
+    return f;
+);
+
+updatePackage = method();
+updatePackage String := packagename -> (
+    needsPackage packagename;
+    ImportsExports := sort unique((value packagename).Options.PackageImports | (value packagename).Options.PackageExports);
+    dismiss packagename;
+    UpdateDone := false;
+    v := null;
+    for P in ImportsExports do (
+        try v = updatePackageFromDevelopment(P,auxFiles P) 
+        then (if v =!= null then UpdateDone = true) 
+        else <<"--warning: this function is not yet available for the dependency on package "|packagename<<endl
+    );
+    v = updatePackageFromDevelopment(packagename,auxFiles packagename);
+    if (not UpdateDone) and v === null then return;
+    <<endl<<endl;
+    e := "";
+    while not(e == "y" or e == "yes" or e == "Y" or e == "Yes") do (
+        e = read("Do you want to restart Macaulay2 now (y/n) ");
+        if e == "n" or e == "no" or e == "N" or e == "No" then return;
+    );
+    return restart;
```

Comment:
Don't bother with this -- users know how to restart.

Reply:
It seemed nice to me like that. It is also a way of pointing out that a restart is required.

Reply:
Maybe just point that out in a message.  Offering to do it, however, seems too intrusive, and it's giving a job to the user : answer this question.  The user might be in a hurry.